xds/internal/xdsclient/transport/transport.go

```diff
@@ -0,0 +1,564 @@
+// of processing this request, it is queued and will be sent out once a valid
+// stream exists.
+//
+// If a successful response is received, the update handler callback provided at
+// creation time is invoked. If an error is encountered, the stream error
+// handler callback provided at creation time is invoked.
+func (t *Transport) SendRequest(url string, resources []string) {
+	t.adsRequestCh.Put(&resourceRequest{
+		url:       url,
+		resources: resources,
+	})
+}
+
+func (t *Transport) newAggregatedDiscoveryServiceStream(ctx context.Context, cc *grpc.ClientConn) (adsStream, error) {
+	// The transport retries the stream with an exponential backoff whenever the
+	// stream breaks. But if the channel is broken, we don't want the backoff
+	// logic to continuously retry the stream. Setting WaitForReady() blocks the
+	// stream creation until the channel is READY.
+	//
+	// TODO(easwars): Make changes required to comply with A57:
+	// https://github.com/grpc/proposal/blob/master/A57-xds-client-failure-mode-behavior.md
+	return v3adsgrpc.NewAggregatedDiscoveryServiceClient(cc).StreamAggregatedResources(ctx, grpc.WaitForReady(true))
+}
+
+func (t *Transport) sendAggregatedDiscoveryServiceRequest(stream adsStream, resourceNames []string, resourceURL, version, nonce, errMsg string) error {
+	req := &v3discoverypb.DiscoveryRequest{
+		Node:          t.nodeProto,
+		TypeUrl:       resourceURL,
+		ResourceNames: resourceNames,
+		VersionInfo:   version,
+		ResponseNonce: nonce,
+	}
+	if errMsg != "" {
+		req.ErrorDetail = &statuspb.Status{
+			Code: int32(codes.InvalidArgument), Message: errMsg,
+		}
+	}
+	if err := stream.Send(req); err != nil {
+		return fmt.Errorf("sending ADS request %s failed: %v", pretty.ToJSON(req), err)
+	}
+	t.logger.Debugf("ADS request sent: %v", pretty.ToJSON(req))
+	return nil
+}
+
+func (t *Transport) recvAggregatedDiscoveryServiceResponse(stream adsStream) (resources []*anypb.Any, resourceURL, version, nonce string, err error) {
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, "", "", "", fmt.Errorf("failed to read ADS response: %v", err)
+	}
+	t.logger.Infof("ADS response received, type: %v", resp.GetTypeUrl())
+	t.logger.Debugf("ADS response received: %v", pretty.ToJSON(resp))
+	return resp.GetResources(), resp.GetTypeUrl(), resp.GetVersionInfo(), resp.GetNonce(), nil
+}
+
+// run starts an ADS stream (and backs off exponentially, if the previous
+// stream failed without receiving a single reply) and runs the sender and
+// receiver routines to send and receive data from the stream respectively.
+func (t *Transport) run(ctx context.Context) {
+	go t.send(ctx)
+	// TODO: start a goroutine monitoring ClientConn's connectivity state, and
+	// report error (and log) when stats is transient failure.
+
+	retries := 0
+	lastStreamStartTime := time.Time{}
+	for ctx.Err() == nil {
+		dur := time.Until(lastStreamStartTime.Add(t.backoff(retries)))
```

Comment:
(As above)

Reply:
Done. Yes, I think we can put this functionality in a helper function.
Filed https://github.com/grpc/grpc-go/issues/5813.